File:  max_depth_of_binary_tree/max_depth_of_binary_tree.rb
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

class Solution
  def max_depth(root)
    return 0 if root.nil?
    1 + [max_depth(root.left), max_depth(root.right)].max
  end
end

require 'rspec'

RSpec.describe Solution do
  let(:solution) { Solution.new }

  it 'calculates max depth correctly' do
    root = TreeNode.new(3)
    root.left = TreeNode.new(9)
    root.right = TreeNode.new(20, TreeNode.new(15), TreeNode.new(7))

    expect(solution.max_depth(root)).to eq(3)
  end

  it 'returns 0 for an empty tree' do
    expect(solution.max_depth(nil)).to eq(0)
  end

  # it 'returns 1 for a single node' do
  #   root = TreeNode.new(1)
  #   expect(solution.max_depth(root)).to eq(1)
  # end
end

File:  max_depth_of_binary_tree/max_depth_of_binary_tree.go
package main

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    leftDepth := maxDepth(root.Left)
    rightDepth := maxDepth(root.Right)
    return 1 + max(leftDepth, rightDepth)
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

package main

import "testing"

func TestMaxDepth(t *testing.T) {
    root := &TreeNode{Val: 3}
    root.Left = &TreeNode{Val: 9}
    root.Right = &TreeNode{Val: 20, Left: &TreeNode{Val: 15}, Right: &TreeNode{Val: 7}}

    if maxDepth(root) != 3 {
        t.Errorf("Expected 3, got %d", maxDepth(root))
    }

    if maxDepth(nil) != 0 {
        t.Errorf("Expected 0, got %d", maxDepth(nil))
    }

    singleNode := &TreeNode{Val: 1}
    if maxDepth(singleNode) != 1 {
        t.Errorf("Expected 1, got %d", maxDepth(singleNode))
    }
}
